### ISSUE TITLE
Segfault with Provider

### ISSUE THREAD

**Issue by Myushu** [google/fruit#16]
Hi,

I have a segfault when a tried to use `fruit::Provider`.
I have many differents writer and i need to choose one at runtime.
In the following code I have only one `Writer,` but it's enough to cause the segfault.
Did i do  something wrong ?

**output:**
```
Writer::write test1
[1]    90298 segmentation fault  ./toto
```

**code :**
```
#include <fruit/fruit.h>
#include <iostream>

class Writer
{
public:
  virtual void write(std::string a) const = 0;
};

class WriterImpl : public Writer
{
public:
  INJECT(WriterImpl()) = default;
  void write(std::string a) const override { std::cout << "Writer::write " << a << std::endl; }
};

class Greeter
{
public:
  virtual void greet() const = 0;
};

class GreeterImpl : public Greeter
{
public:
  INJECT(GreeterImpl(fruit::Provider<Writer> writerProvider))
    : _writer(writerProvider.get())
  {
    _writer->write("test1");
  }
  void greet() const override
  {
    std::cout << "Greeter::greet" << std::endl;
    _writer->write("test2");
  }

private:
  Writer *_writer;
};

int main()
{
  fruit::Injector<Greeter> injector(fruit::createComponent()
    .bind<Writer, WriterImpl>()
    .bind<Greeter, GreeterImpl>()
  );
  Greeter *greeter(injector);
  greeter->greet();
  return 0;
}

```

**Comment by poletti-marco** [comment]
Thanks a lot for the report, calling .get() on a Provider within a constructor is quite a corner case (if you want a Writer* to always be injected, you could just inject Writer* directly) but it most definitely shouldn't cause undefined behavior. I'll release a bugfix release soon.

**Comment by poletti-marco** [comment]
I've now released Fruit 2.0.4 with the fix for this issue. Please update to get the fix, and thanks again for reporting this issue.
